readme.md

tool for interactively viewing climate priming maps


INSTALLATION
------------

install numpy, matplotlib, gdal, pyyaml, multigrids,  bokeh

install spicebox (https://github.com/rwspicer/spicebox)



RUNNING
-------
set data_root in data_sources.py

run: bokeh serve main.py

go to: http://localhost:5006/main. The page will take a while to load data 
the first time.


Running with docker
-------------------

Install Docker

set up the path to data in the volumes of the docker-compose file

run: `docker compose up`

go to:  http://localhost:8000/main. The page will take a while to load data 
the first time.





changelog.md
# CHANGLOG

## [0.4.0] 2021-01-06
### Added
- Docker support

## [0.3.1] 2020-12-10
### Changed
- path to data in data_sources.py 

## [0.3.0] 2020-12-10
### Added 
- added caching of multigrid views
- ACP site
- Toggle for viewing met as % difference from mean
- met_filter function
- display of "clicked" value on map
- buttons for previous/next year

### Updated
- CP_MIN and CP_MAX are -50 and 50 respectively 
- cp_map plot settings
- Utqiagvik site metadata

## [0.2.0] 2020-12-02
### Added
- climate priming threshold overlay
- thermokarst predisposition model overly
- added options for loading data from multigrid objects

## [0.1.0] 2020-11-21
### Added 
- Initial Version
- CP map
- CP average plot
- year slider
- met maps
- region drop down
- readme


maps.py

from bokeh.plotting import figure

import constants,colors

from bokeh.models import LinearColorMapper, ColorBar

from bokeh.events import Tap

def create_cp_map(data):
    TOOLTIPS = [
        ("(x,y)", "($x, $y)"),
        # ("(x map, y map)", "(@x_geo{0,0.000}, @y_geo{0,0.000})"),
        # ("(lat, long)", "(@lat{0,0.000}, @long{0,0.000})"),
        ("value", "@cp_vals{0,0.000}")
    ]
    img_shape = data.data['cp'][0].shape
    p = figure(
        tooltips= TOOLTIPS,
        plot_height = img_shape[0] * constants.WIDTH*2//img_shape[1],
        plot_width = constants.WIDTH *2 ,
        tools=[],
        title="Utqiaġvik and Point Barrow"
        )

    p.toolbar.logo = None
    p.toolbar_location = None
    # p.match_aspect = True
    p.x_range.range_padding = p.y_range.range_padding = 0
    # p.sizing_mode = 'scale_height'
    # must give a vector of image data for image parameter

    cmapper = LinearColorMapper(
        colors.blue_red , 
        low=constants.CP_MIN, high=constants.CP_MAX, 
        nan_color='white',
        low_color="gray"
    )



    p.image(
        image='cp', x=0, y=0, dw=10, dh=10, 
        color_mapper=cmapper,
        # level="image",
        source=data,
    )
    # p.grid.grid_line_width = 0.5
    p.xgrid.grid_line_color = None
    p.ygrid.grid_line_color = None
    p.axis.visible = False

   

    return p



def create_met_map(data, met_name, offset, color, c_min, c_max):
    TOOLTIPS = [
        ("(x,y)", "($x, $y)"),
        # ("(x map, y map)", "(@x_geo{0,0.000}, @y_geo{0,0.000})"),
        # ("(lat, long)", "(@lat{0,0.000}, @long{0,0.000})"),
        ("value", "@"+met_name+"{0,0.000}")
    ]
    met_map = figure(
        tooltips= TOOLTIPS,
        plot_height = 200,
        plot_width = 200 ,
        tools=[],
        title=""
        )
    
    cmapper = LinearColorMapper(
        color, 
        low=c_min, high=c_max, 
        nan_color='white'
    )

    met_map.image(
        image=met_name, x=0, y=0, dw=10, dh=10, 
        color_mapper=cmapper,
        # level="image",
        source=data,
    )
    # p.grid.grid_line_width = 0.5
    met_map.xgrid.grid_line_color = None
    met_map.ygrid.grid_line_color = None
    met_map.axis.visible = False

    met_map.toolbar.logo = None
    met_map.toolbar_location = None
    met_map.x_range.range_padding = met_map.y_range.range_padding = 0

    color_bar = ColorBar(color_mapper=cmapper, name='cb', border_line_color=None, location=(0,0), width=10)

    met_map.add_layout(color_bar, 'right')

    return met_map


plots.py
import numpy as np



from bokeh.plotting import figure
from bokeh.transform import linear_cmap


import constants, colors



def create_average_plot(current_display, timeseries, start_year):

    

    average_plot = figure(
        plot_width= constants.WIDTH*4, plot_height= constants.WIDTH+100, 
        x_range=(1900, 2016), y_range=( constants.CP_MIN,  constants.CP_MAX), 
        tools=[],
        title="Yearly Average for Area"
    )


    average = timeseries.data['average']
    ## 1901-1950 average_line
    average_plot.ray(
        x=[start_year-1], y=[np.nanmean(average[:1950-start_year])], 
        length=0, angle=0, line_width=3, alpha=0.5, color="blue",
        legend_label="1901-1950 regional average"
    )


    color_mapper = linear_cmap(
        field_name='average',palette=colors.blue_red, 
        low=constants.CP_MIN, high=constants.CP_MAX
    )
    average_plot.circle(
        'years', 'average', size=10, color=color_mapper,
        alpha=1, source = timeseries
    )
    
    average_plot.circle(
        x='current_year', y='current_average', size=20, color="green", 
        alpha=.5, source=current_display,
        legend_label="Current Year"
    )

    average_plot.legend.location = "bottom_left"
    average_plot.legend.click_policy="hide"

    return average_plot


sites.py


LOCATIONS = {
    "Utqiaġvik": {
        "geolocation": (70.9, -156),
        "name": "Utqiaġvik and Point Barrow",
        "papers": [""], 
        "data": {},
        "description": """
            <p>BW start of degradation in early  1950s mostly stable by early 2000s  
            (~90% IW studied stable) Kanesvskiy et al. (2017) </p>
    """
    },
    "Prudhoe Bay": {
        "geolocation": (70.2288, -148.4186), ## location of Jorgenson et al. (2015)
        "name": "Prudhoe Bay",
        "data": {},
        "papers": ["Raynolds et al. (2014)", "Jorgenson et al. (2015)" ], 
        "description": """
            367 ha in 1968 -> 663 ha in 2010 with most of change after 1990 -- Raynolds et al. (2014).
            Thermokarst extent was relatively small from 1949 (0.9%) to 1988 
            (1.5%), abruptly increased from 1988 to 2004 (6.3%), and increased 
            slightly from 2004 to 2012 (7.5%). Given that ice wedges occupy 
            32% of the upper 1 m of permafrost in the study area based on 
            the geometric method, we estimate that ~23% (7.5%/32%) 
            of the surface of the ice wedges has been degraded, but the 
            degradation was usually limited to the top ~0.5 m of the wedges. 
            -- Jorgenson et al. (2015)  
            """
    },
    "NPRA": {
        "geolocation": (70.25, -151.5), ## location of Jorgenson et al. (2015)
        "name": "NPRA near Nuiqsut",
        "data": {},
        "papers": ["Jorgenson et al. (2003)", ], 
        "description": """
            degradation thought to be due to warmer temperature in 1989 - 1998 
            --Jorgenson et al. (2003)
            """
    },
    "ACP": {
        "geolocation": (70.25, -151.5), ## location of Jorgenson et al. (2015)
        "name": "Arctic Coastal Plain",
        "data": {},
        "papers": [ ], 
        "description": """"""
    }
}




data_sources.py
from multigrids import tools, TemporalGrid
import glob
import os
import numpy as np
from spicebox import raster, transforms

from sites import LOCATIONS

data_root = '../../data'

DATA_TYPE = "multigrid"

REGION = 'ACP'



path_dict = {
   "ewp": os.path.join(data_root, 'V1/precipitation/early-winter', REGION, 'v3/cru/', DATA_TYPE),
   "fwp": os.path.join(data_root, 'V1/precipitation/full-winter', REGION, 'v3/cru/', DATA_TYPE),
   "tdd": os.path.join(data_root, 'V1/degree-day/thawing', REGION, 'v4/cru/', DATA_TYPE),
   "fdd": os.path.join(data_root, 'V1/degree-day/freezing', REGION, 'v4/cru/', DATA_TYPE),
   "cp":  os.path.join(data_root, 'V1/thermokarst/initiation-regions', REGION, 'v5-d/PDM-5var/without_predisp/', DATA_TYPE)
}

view_path = os.path.join(data_root, 'V1', 'cp-explorer-views')
try: 
    os.makedirs (view_path)
except:
    pass

import tempfile
tempfile.tempdir = os.path.join(data_root,'tmp')

predisp_model, pm_md = raster.load_raster(
    os.path.join(
        data_root, 
        'V1/thermokarst/predisposition-model/ACP/v2',
        'ACP-tk-predisp-model.tif'
    )
)


def load_dataset(dataset_name, path, sites, start_timestep):

    if DATA_TYPE == "tiff":

        files = glob.glob(os.path.join(path, "*.tif"))

        load_params = {
            "method": "tiff",
            "directory": path
        }
        create_params = {
            'start_timestep': start_timestep,
            'raster_metadata': tools.get_raster_metadata(files[0])
        }

        full_data = tools.load_and_create(load_params, create_params)
    else:
        files = glob.glob(os.path.join(path, "*.yml"))
        full_data = TemporalGrid(files[0])
    
    # data_by_site = {}
    for site in sites:
        
        if site == "ACP":
            start_year = full_data.config['start_timestep']

            grid_shape = full_data.config['grid_shape']
            full_data.config['mean'] = np.nanmean(full_data.grids[:1951-start_year],axis=0).reshape(grid_shape)
            sites[site]['data'][dataset_name] = full_data
            pdm = predisp_model
            print 


        else:
            current_vp = os.path.join(view_path, '%s-%s.yml' % (site, dataset_name))
            try:
                view = TemporalGrid(current_vp)
                print('view loaded:', current_vp )
            except FileNotFoundError as e: 
                print (e)
                view = full_data.zoom_to(
                    sites[site]['geolocation'], location_format="WGS84"
                )
                start_year = view.config['start_timestep']
                grid_shape = view.config['grid_shape']
                # print(view.grids[:1951-start_year].mean(axis=0))
                view.config['mean'] = np.nanmean(view.grids[:1951-start_year],axis=0).reshape(grid_shape)
                # print(view.config['mean'])

                
                view.save(current_vp)

            

            sites[site]['data'][dataset_name] = view
            geo = transforms.from_wgs84(
                sites[site]['geolocation'], pm_md['projection']
            )
            pixel = transforms.to_pixel(geo, full_data.config['raster_metadata']['transform']).astype(int)
            pdm = raster.zoom_to(predisp_model, pixel) 

        pdm[pdm<0] = -np.inf
        pdm = pdm / 100

        sites[site]['predisp_model'] = pdm[::-1]
    return sites


def load_all(path_dict, sites, start_timestep=1901):

    sites = LOCATIONS
    for dataset in ["fdd", "tdd", "ewp", "fwp", "cp"]:
        sites = load_dataset(
            dataset , path_dict[dataset], sites, start_timestep
        )

    return sites

GLOBAL_SITE_DATA = load_all(path_dict, LOCATIONS, 1901)

def create_data(site_data, init_year = 1902):

    # # print('a')
    # to_x_geo = lambda x, y, gt: gt[0] + (x + .5)*gt[1] + (y + .5)*gt[2]
    # to_y_geo = lambda x, y, gt: gt[3] + (x + .5)*gt[4] + (y + .5)*gt[5]
    # md =  multigrid.config['raster metadata']
    

    # in_crs = CRS.from_wkt(md.projection)
    # gps = CRS.from_epsg(4326)
    # transformer = Transformer.from_crs(in_crs, gps)
    # # print('b')
    # x = np.zeros(init.shape)
    # y = np.zeros(init.shape)
    # x[:] = range(init.shape[1])
    # y.T[:] = range(init.shape[0])

    # # print('c')

    # x_geo = to_x_geo(x,y,md.transform)
    # y_geo = to_y_geo(x,y,md.transform)
    # print('d')
    data = dict(
        cp=[site_data['data']['cp'][init_year][::-1]], 
        cp_vals=[site_data['data']['cp'][init_year][::-1]],
        # x_geo = [x_geo[::-1]],
        # y_geo = [y_geo[::-1]],
        # lat = [transformer.transform(x_geo,y_geo)[0] [::-1]],
        # long = [transformer.transform(x_geo,y_geo)[1] [::-1]],
        # raster_metadata = [md]
    )
    # print('e')

    data['current_year'] = [init_year]
    data['current_average'] = [0]
    data['tdd_last'] = [site_data['data']['fdd'][init_year - 1][::-1]]
    data['tdd'] = [site_data['data']['fdd'][init_year][::-1]]
    data['fdd'] = [site_data['data']['fdd'][init_year - 1][::-1]]
    data['ewp'] = [site_data['data']['ewp'][init_year - 1][::-1]]
    data['fwp'] = [site_data['data']['fwp'][init_year - 1][::-1]]
    data['predisp'] = [predisp_model]

    return data

def create_area_average_data(multigrid):

    average = []
    for yr in range(multigrid.config['num_timesteps']):
        # print(multigrid.grids[yr])
        average.append(np.nanmean(multigrid.grids[yr]))
    return average


fdd-explorer.py
from multigrids import MultiGrid, tools
import os
import numpy as np
import matplotlib.pyplot as plt

from bokeh.io import curdoc
from bokeh.layouts import column, row, layout
from bokeh.plotting import figure, output_file, show
from bokeh import palettes
from bokeh.transform import linear_cmap

from bokeh.models import TextInput, ColumnDataSource, FileInput, Div, Dropdown, Slider,ColorBar, LinearColorMapper


## TEST DATA setup
path = "/Volumes/scarif/data/V2/datasets/derived/fdd_ACP_CRU_v3/tiff/"
load_params = {
    "method": "tiff",
    "directory": path
}
create_params = {
    'start_timestep':1901,
    'raster_metadata': tools.get_raster_metadata(os.path.join(path, "freezing_degree-day_1901.tif"))
}

test_data = tools.load_and_create(load_params, create_params)

UTQ = test_data.zoom_to((70.9, -156), location_format="WGS84")
PRU = test_data.zoom_to((70, -151), location_format="WGS84")

print (UTQ.config['grid_shape'])
print (PRU.config['grid_shape'])


## TEST DATA setup

WIDTH = 200

palette_opts = {
    "viridis": palettes.Viridis[256],
    "magma": palettes.Magma[256]
}

def create_data(multigrid, year):

    # # print('a')
    # to_x_geo = lambda x, y, gt: gt[0] + (x + .5)*gt[1] + (y + .5)*gt[2]
    # to_y_geo = lambda x, y, gt: gt[3] + (x + .5)*gt[4] + (y + .5)*gt[5]
    # md =  multigrid.config['raster metadata']
    init = multigrid[year]

    # in_crs = CRS.from_wkt(md.projection)
    # gps = CRS.from_epsg(4326)
    # transformer = Transformer.from_crs(in_crs, gps)
    # # print('b')
    # x = np.zeros(init.shape)
    # y = np.zeros(init.shape)
    # x[:] = range(init.shape[1])
    # y.T[:] = range(init.shape[0])

    # # print('c')

    # x_geo = to_x_geo(x,y,md.transform)
    # y_geo = to_y_geo(x,y,md.transform)
    # print('d')
    data = dict(
        image=[init[::-1]], 
        # x_geo = [x_geo[::-1]],
        # y_geo = [y_geo[::-1]],
        # lat = [transformer.transform(x_geo,y_geo)[0] [::-1]],
        # long = [transformer.transform(x_geo,y_geo)[1] [::-1]],
        # raster_metadata = [md]
    )
    # print('e')
    return data

def create_figure(data):
    TOOLTIPS = [
        ("(x,y)", "($x, $y)"),
        # ("(x map, y map)", "(@x_geo{0,0.000}, @y_geo{0,0.000})"),
        # ("(lat, long)", "(@lat{0,0.000}, @long{0,0.000})"),
        ("value", "@image{0,0.000}")
    ]
    img_shape = data.data['image'][0].shape
    p = figure(
        tooltips= TOOLTIPS,
        plot_height = img_shape[0] * WIDTH*2//img_shape[1],
        plot_width = WIDTH *2 ,
        tools=[],
        title="Utqiaġvik and Point Barrow"
        )

    p.toolbar.logo = None
    p.toolbar_location = None
    p.x_range.range_padding = p.y_range.range_padding = 0
    # p.sizing_mode = 'scale_height'
    # must give a vector of image data for image parameter

    cmapper = LinearColorMapper(palettes.Viridis[256], low=-6000, high=-3500, nan_color='white')

    p.image(
        image='image', x=0, y=0, dw=10, dh=10, 
        color_mapper=cmapper,
        # palette=palettes.linear_palette(palette_opts[data.data['current_palette'][0]],data.data['current_colors'][0] ), 
        level="image",
        source=data,
    )
    # p.grid.grid_line_width = 0.5
    p.xgrid.grid_line_color = None
    p.ygrid.grid_line_color = None
    p.axis.visible = False

    return p


def create_area_average_data(multigrid):

    average = []
    for yr in range(multigrid.config['num_timesteps']):
        # print(multigrid.grids[yr])
        average.append(np.nanmean(multigrid.grids[yr]))
    return average


def app():

    start_year = UTQ.config['start_timestep']
    end_year = start_year + UTQ.config['num_timesteps']
    years = range(start_year,end_year)
    
    current_palette = 'viridis'
    current_colors = 256
    display_year = start_year
    # print(display_year)

    display_region = ColumnDataSource()
    display_region.data = create_data(UTQ, display_year )
    display_region.data['current_palette'] = [current_palette]
    display_region.data['current_colors'] = [current_colors]
    display_region.data['current_year'] = [display_year ]
    cp_map = create_figure(display_region)

    average = create_area_average_data(UTQ)
    # print(average)


    average_plot = figure(
        plot_width=WIDTH*4, plot_height=WIDTH, 
        x_range=(1900, 2016), y_range=(-6000, -3500), tools=[],
        title="Yearly Average for Area"
    )

    ## 1901-1950 average_line
    average_plot.ray(x=[start_year-1], y=[np.nanmean(average[:1950-start_year])], length=0, angle=0, line_width=3, alpha=0.5, color="blue",legend_label="1901-1950 regional average")
    # average_plot.ray(x=[start_year], y=[np.nanmean(average[:1950-start_year])], length=0, angle=np.pi, line_width=3,alpha=0.5, color="blue", legend_label="1901-1950 regional average")


    timeseries = ColumnDataSource()
    color_mapper = linear_cmap(field_name='average',palette=palettes.Viridis[256], low=-6000, high=-3500)
    timeseries.data = {"years": years, "average": average}
    average_plot.circle('years', 'average', size=10, color=color_mapper , alpha=0.75, source = timeseries)
    

    display_average = average[display_year-start_year]

    display_region.data['current_average'] = [display_average ]
    average_plot.circle(x='current_year', y='current_average', size=20, color="red", alpha=.5, source=display_region )

    average_plot.legend.location = "top_left"
    average_plot.legend.click_policy="hide"

    hidden_plot = figure(
        plot_width=100, plot_height=WIDTH*2, 
        tools=[],
        title = "Degree-days",title_location="right",
        toolbar_location=None, min_border=0, 
        outline_line_color=None
    )
    hidden_plot.toolbar.logo = None
    hidden_plot.toolbar_location = None
    hidden_plot.title.align="center"
    hidden_plot.title.text_font_size = '12pt'

    color_bar = ColorBar(color_mapper=color_mapper['transform'], width=8,  location=(0,0))
    hidden_plot.add_layout(color_bar, 'right')
    


    year_slider = Slider(
        start=start_year, end=end_year, value=display_year, step=1, title="Year"
    )
    def change_display_year(attrname, old, new):
       
        display_region.data['image']  = [UTQ[int(new)][::-1]]
        display_average = average[new-start_year]
        display_region.data['current_average'] = [display_average ]
        display_region.data['current_year'] = [new]

       
    year_slider.on_change('value_throttled', change_display_year)

    inputs = column(year_slider)

    title = Div(text="<H1>FREEZING DEGREE-DAY EXPLORER<H1>", width=1000)
    
    curdoc().add_root(
        layout([
            [title],
            [inputs, cp_map, hidden_plot],
            [average_plot]
        ]) 
        
    )
    curdoc().title = "Freezing Degree-day Priming Explorer"

    

    

# if __name__=="__main__":
app() 


main.py
from multigrids import MultiGrid, tools

from spicebox import transforms
import os
import numpy as np
import matplotlib.pyplot as plt

from bokeh.io import curdoc
from bokeh.layouts import column, row, layout, Spacer
from bokeh.plotting import figure, output_file, show
from bokeh import palettes
from bokeh.transform import linear_cmap

from bokeh.models import TextInput, ColumnDataSource, FileInput, Div, Dropdown, Slider,ColorBar, LinearColorMapper, Title, Toggle, Button
from bokeh.events import Tap

from copy import deepcopy
## TEST DATA setup
# path = "/Volumes/scarif/data/V2/datasets/derived/fdd_ACP_CRU_v3/tiff/"
# load_params = {
#     "method": "tiff",
#     "directory": path
# }
# create_params = {
#     'start_timestep':1901,
#     'raster_metadata': tools.get_raster_metadata(os.path.join(path, "freezing_degree-day_1901.tif"))
# }

# test_data = tools.load_and_create(load_params, create_params)

# UTQ = test_data.zoom_to((70.9, -156), location_format="WGS84")
# PRU = test_data.zoom_to((70, -151), location_format="WGS84")

# print (UTQ.config['grid_shape'])
# print (PRU.config['grid_shape'])


## TEST DATA setup












import plots, constants, maps, colors
import data_sources

# from sites import LOCATIONS


def met_filter(data, year, toggle):
    if toggle:
        mean = data.config['mean']
        grid = data[year]

        diff = ( (grid - mean )/ np.abs(mean) ) * 100

        print("mean", mean[-1,-1], 'val', grid[-1,-1], diff[-1,-1])


        return diff[::-1]
    return data[year][::-1]

def update_display(sites, maps, timeseries, values):

    cp = sites[values['region']]['data']['cp']
    start_year = cp.config['start_timestep']
    end_year = start_year + cp.config['num_timesteps']

    timeseries.data['average'] = data_sources.create_area_average_data(cp)
    display_average = timeseries.data['average'][ values['year'] -start_year ]
    maps.data['current_average'] = [display_average ]
    maps.data['current_year'] = [values['year']]

    pd_filter = 1
    if values['predisp']['toggle']: 
        pd_filter = sites[values['region']]['predisp_model']
    
    cp_map = \
        sites[values['region']]['data']['cp'][values['year']][::-1] * pd_filter

    maps.data['cp_vals'] = [deepcopy(cp_map)]

    if values['threshold']['toggle']:
        cp_map = deepcopy(cp_map)  # Ensures data is not over written ## check if needed?
        cp_map[cp_map >= values['threshold']['value']] = constants.CP_MAX - 1 
        cp_map[cp_map < values['threshold']['value']] = constants.CP_MIN + 1 
    
    
    cp_map[np.isinf(cp_map)] = constants.CP_MIN - 10    
    maps.data['cp'] = [cp_map]
    
    

    for m_var in ['tdd_last', 'fdd', 'ewp', 'fwp']:
        maps.data[m_var] = [
            met_filter(
                sites[values['region']]['data'][m_var.replace('_last','')],
                values['year'] - 1, 
                values['met']['toggle']
            )
        ]
    maps.data['tdd'] = [
        met_filter(
            sites[values['region']]['data']['tdd'], 
            values['year'], 
            values['met']['toggle']
        )
    ]
       

    





loc_div_txt = """
<p> location:  %fN %fW </p>
<p> value:  %f </p>
"""

def app():

    sites = data_sources.GLOBAL_SITE_DATA

    # cp_threshold_toggle = False
    # cp_threshold = 0

    # predisp_toggle = False
    # met_toggle = False

    init_region = "Utqiaġvik"
    init = sites[init_region]['data']['cp']
    start_year = init.config['start_timestep']
    end_year = start_year + init.config['num_timesteps']
    
    current_values = {
        'region': init_region,
        'year': start_year + 1,
        'met': {'toggle': False},
        'threshold': {'toggle': False, 'value':0},
        'predisp': {'toggle': False},
    }

    
    
    display = ColumnDataSource()
    display.data = data_sources.create_data(
        sites[current_values['region']], 
        current_values['year']
    )

    description = Div(text="<h2>Notes for area:</h2><p>"+ sites[ current_values['region'] ]['description'] + "</p>")

    timeseries = ColumnDataSource()
    timeseries.data = {
        "years": range( start_year, end_year ), 
        "average": data_sources.create_area_average_data(
            sites[ current_values['region'] ]['data']['cp']
        )
    }

    # current_year =current_values['year']

    display_average = timeseries.data['average'][current_values['year'] - start_year]

    # display.data['current_average'] = [display_average]
    # display.data['region'] = [current_region ]
    
    # display.data['cp_threshold_toggle'] = [cp_threshold_toggle]
    # display.data['cp_threshold'] = [cp_threshold]
    # display.data['predisp_toggle'] = [predisp_toggle]
    # display.data['met_toggle'] = [met_toggle]

    update_display(sites, display, timeseries, current_values)
    
    cp_map = maps.create_cp_map(display)

    click_val = Div(text=loc_div_txt % (0, 0 , 0))
   
    def click_clack(event):

        current_region = current_values['region']

        current_year =current_values['year']
        predisp_toggle = current_values['predisp']['toggle']

        multiplier = 1
        if predisp_toggle: 
            multiplier = sites[current_region]['predisp_model']

        cur_map = sites[current_region]['data']['cp'][current_year][::-1] * multiplier

        col = int((event.x * cur_map.shape[1]) / 10) 
        row = int((event.y * cur_map.shape[0]) / 10) 

        md = sites[current_region]['data']['cp'].config['raster_metadata']

        val = cur_map[row, col]

        wgs_84 = transforms.to_wgs84(transforms.to_geo([row, col], md['transform']), md['projection'])

        click_val.text = loc_div_txt % (wgs_84[0], abs(wgs_84[1]) , val)

        # print(col, row, val)

    cp_map.on_event(Tap, click_clack)

    average_plot = plots.create_average_plot(display, timeseries, start_year)

    


    ### COLOR BAR
    hidden_plot = figure(
        plot_width=100, plot_height=constants.WIDTH*2, 
        tools=[],
        title = "CLIMATE PRIMING VAL",title_location="right",
        toolbar_location=None, min_border=0, 
        outline_line_color=None
    )
    hidden_plot.toolbar.logo = None
    hidden_plot.toolbar_location = None
    hidden_plot.title.align="center"
    hidden_plot.title.text_font_size = '12pt'

    color_mapper = linear_cmap(
        field_name='average',palette=colors.blue_red, 
        low=constants.CP_MIN, high=constants.CP_MAX
    )
    color_bar = ColorBar(color_mapper=color_mapper['transform'], width=8,  location=(0,0))
    hidden_plot.add_layout(color_bar, 'right')
    

    current_year  =current_values['year']
    last_year = str(current_year - 1)
    current_year = str(current_year)
    
    
    ### MET MAPS
    tdd_last_map = maps.create_met_map(
        display, "tdd_last", 1,
        palettes.Reds256[::-1], constants.TDD_MIN, constants.TDD_MAX
    )
    tdd_last_map.add_layout(Title(text=last_year, name='title'), 'above')
    tdd_last_map.add_layout(Title(text="Thawing Degree Days "), 'above')

    fdd_map = maps.create_met_map(display, "fdd", 1,
        palettes.Blues256, constants.FDD_MIN, constants.FDD_MAX
    )
    fdd_map.add_layout(Title(text=last_year + '-' + current_year, name='title'), 'above')
    fdd_map.add_layout(Title(text="Freezing Degree Days "), 'above')

    tdd_map = maps.create_met_map(
        display, "tdd", 0,
        palettes.Reds256[::-1], constants.TDD_MIN, constants.TDD_MAX
    )
    tdd_map.add_layout(Title(text=current_year, name='title'), 'above')
    tdd_map.add_layout(Title(text="Thawing Degree Days "), 'above')

    ewp_map = maps.create_met_map(
        display, "ewp", 1,
        palettes.Greens256[::-1], constants.EWP_MIN, constants.EWP_MAX
    )
    ewp_map.add_layout(Title(text="Oct "+ last_year+" - Nov " + last_year, name='title'), 'above')
    ewp_map.add_layout(Title(text="Early Winter Precipitation [mm]"), 'above')

    fwp_map = maps.create_met_map(
        display, "fwp", 1,
        palettes.Greens256[::-1], constants.FWP_MIN, constants.FWP_MAX
    )
    fwp_map.add_layout(Title(text="Oct "+ last_year +" - Mar " + current_year, name='title'), 'above')
    fwp_map.add_layout(Title(text="Total Winter Precipitation [mm]"), 'above')

    # fwp_map.add_layout() Div(text="<div>Total winter precipitation<div></div>  + "</div>")


    ##### CONTROLS ---------

    options = [(sites[r]['name'], r) for r in sites]
    region_dropdown = Dropdown(label="region", menu=options)

    year_slider = Slider(
        start=start_year+1, end=end_year - 1, 
        value=current_values['year'], step=1, title="Year"
    )
    year_next = Button(label="Next Year", width=145)
    year_previous = Button(label="Previous Year", width=145)

    threshold_toggle = Toggle(label="climate priming threshold off")
    threshold_slider = Slider(
        start=constants.CP_MIN, end=constants.CP_MAX, 
        step = .5,
        value=current_values['threshold']['value'],
        title="Climate Priming Threshold"
    )

    predisp_toggle_button = Toggle(label="predisposition model overlay off")

    met_toggle_button = Toggle(label="Show Met as difference from mean off")

    ##### CALLBACKS

    def change_display_year(attrname, old, new):
     
        current_values['year'] = new
  

        update_display(
            sites, 
            display, timeseries, 
            current_values
        )

        last_year = str(current_values['year'] - 1)
        year = str(current_values['year'])

        tdd_last_map.select(name="title").text = last_year
        fdd_map.select(name="title").text = last_year + '-' + year
        tdd_map.select(name="title").text = year
        ewp_map.select(name="title").text = "Oct "+ last_year+" - Nov " + last_year
        fwp_map.select(name="title").text = "Oct "+ last_year +" - Mar " + year


    def next_year(event):

        yr = min(current_values['year'] + 1, end_year)
        if yr == end_year:
            return
        change_display_year("", "", yr)
        year_slider.value = current_values['year']
    
    def previous_year(event):

        yr = max(current_values['year'] - 1, start_year + 1)
        if yr == start_year :
            return
        change_display_year("", "", yr)
        year_slider.value = current_values['year']

    def change_region(event):
       
        current_values['region'] = event.item

        update_display(sites, display, timeseries, current_values)

        description.text = "<h2>Notes for area:</h2><p>"+ \
            sites[current_values['region']]['description'] + "</p>"

    def toggle_threshold (event):

        current_values['threshold']['toggle'] = not current_values['threshold']['toggle']
        update_display(sites, display, timeseries, current_values)

        if current_values['threshold']['toggle']: 
            predisp_toggle_button.label="climate priming threshold on"
        else:
            predisp_toggle_button.label='climate priming threshold off'

    def change_threshold_value(attrname, old, new):

        current_values['threshold']['value'] = new
        if current_values['threshold']['toggle']:
            update_display(sites, display, timeseries, current_values)

        

    def toggle_predisp(event): 

        current_values['predisp']['toggle'] = not current_values['predisp']['toggle']
        update_display(sites, display, timeseries, current_values)

        if current_values['predisp']['toggle']: 
            predisp_toggle_button.label="predisposition model overlay on"
        else:
            predisp_toggle_button.label="predisposition model overlay off"
            
            

    def toggle_met(event): 

        current_values['met']['toggle'] = not current_values['met']['toggle']
        update_display(sites, display, timeseries, current_values)
        
        if current_values['met']['toggle']: 
            met_toggle_button.label="Show Met as difference from mean on"

            diff_c_map = linear_cmap(
                field_name='average',palette=colors.blue_red, 
                low=-25, high=25, nan_color='white'
            )

            tdd_last_map.renderers[0].glyph.color_mapper = diff_c_map['transform']
            fdd_map.renderers[0].glyph.color_mapper = diff_c_map['transform']
            tdd_map.renderers[0].glyph.color_mapper = diff_c_map['transform']
            ewp_map.renderers[0].glyph.color_mapper = diff_c_map['transform']
            fwp_map.renderers[0].glyph.color_mapper = diff_c_map['transform']



            # fdd_map.select('cb')[0].color_mapper = linear_cmap(
            #     field_name='average',palette=colors.blue_red, 
            #     low=-25, high=25, nan_color='white'
            # )['transform']
        else:
            met_toggle_button.label="Show Met as difference from mean off"
            tdd_last_map.renderers[0].glyph.color_mapper = LinearColorMapper(palettes.Reds256[::-1], 
                low=constants.TDD_MIN, high=constants.TDD_MAX, 
                nan_color='white'
            )

            fdd_cmap=LinearColorMapper(palettes.Blues256, 
                low=constants.FDD_MIN, high=constants.FDD_MAX, 
                nan_color='white'
            )
            fdd_map.renderers[0].glyph.color_mapper = fdd_cmap

            tdd_map.renderers[0].glyph.color_mapper = LinearColorMapper(palettes.Reds256[::-1], 
                low=constants.TDD_MIN, high=constants.TDD_MAX, 
                nan_color='white'
            )
            ewp_map.renderers[0].glyph.color_mapper = LinearColorMapper(palettes.Greens256[::-1], 
                low=constants.TDD_MIN, high=constants.TDD_MAX, 
                nan_color='white'
            )
            fwp_map.renderers[0].glyph.color_mapper = LinearColorMapper(palettes.Greens256[::-1], 
                low=constants.TDD_MIN, high=constants.TDD_MAX, 
                nan_color='white'
            )

        
       
    year_slider.on_change('value_throttled', change_display_year)
    region_dropdown.on_click(change_region)
    threshold_toggle.on_click(toggle_threshold)
    threshold_slider.on_change('value_throttled', change_threshold_value)
    predisp_toggle_button.on_click(toggle_predisp)
    year_next.on_click(next_year)
    year_previous.on_click(previous_year)
    met_toggle_button.on_click(toggle_met)


    inputs = column([region_dropdown, year_slider, row([year_previous, year_next] ), predisp_toggle_button, threshold_toggle,threshold_slider, met_toggle_button, click_val])

    title = Div(text="<H1>CLIMATE PRIMING EXPLORER<H1>", width=1000)
    from copy import deepcopy
    dd_row = row([tdd_last_map, fdd_map, tdd_map])
    precip_row = row( [Spacer(width=100, height = 100), ewp_map, fwp_map])
    met_col = column(dd_row, precip_row)

    

    curdoc().add_root(
        layout([
            [title],
            [inputs, cp_map, hidden_plot, met_col  ] ,
            [average_plot, description]
        ]) 
        
    )
    curdoc().title = "Climate Priming Explorer"

    

    

# if __name__=="__main__":
app() 
